### PR TITLE
udp2raw-multiplatform: 20210111.0 (new formula)

### PR DESCRIPTION
**Pull request by crab2313**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/udp2raw-multiplatform.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+class Udp2rawMultiplatform < Formula
+  desc "Multi-platform(cross-platform) version of udp2raw-tunnel client"
+  homepage "https://github.com/wangyu-/udp2raw-multiplatform"
+  url "https://github.com/wangyu-/udp2raw-multiplatform/archive/refs/tags/20210111.0.tar.gz"
+  sha256 "712ad3c79b6ef5bf106c615823d0b0b3865d1c957f9838cf05c23b7ac7024438"
+  license "MIT"
+
+  depends_on "make" => :build
```
Comment (carlocab) [inline Formula/udp2raw-multiplatform.rb:8]: You don't need this.
```suggestion
```

File: Formula/udp2raw-multiplatform.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  depends_on "libnet"
+  depends_on :macos
```
Comment (carlocab) [inline Formula/udp2raw-multiplatform.rb:10]:
```suggestion
```
This doesn't seem accurate. The name is `-multiplatform`, and there are Linux install instructions in their [wiki](https://github.com/wangyu-/udp2raw-multiplatform/wiki/How-to-build#for-linux-debian).

File: Formula/udp2raw-multiplatform.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+  def install
+    ENV["OPT"] = (system_command "libnet-config", args: ["--cflags"]).stdout
```
Comment (carlocab) [inline Formula/udp2raw-multiplatform.rb:13]:
```suggestion
    ENV["OPT"] = Utils.safe_popen_read("libnet-config --cflags")
```
Is this not done by the Makefile?

File: Formula/udp2raw-multiplatform.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    system "make", "mac"
+    bin.install "udp2raw_mp"
+    etc.install "example.conf" => "udp2raw_client.conf"
```
Comment (carlocab) [inline Formula/udp2raw-multiplatform.rb:16]: Is there no `make install`?

File: Formula/udp2raw-multiplatform.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  service do
+    run [opt_bin/"udp2raw_mp", "--conf-file", etc/"udp2raw_client.conf"]
+    keep_alive true
+    require_root true
+    log_path var/"log/udp2raw.log"
+    error_log_path var/"log/udp2raw.log"
+  end
+
+  test do
+    assert_match "udp2raw-tunnel", shell_output("#{bin}/udp2raw_mp -h")
```
Comment (carlocab) [inline Formula/udp2raw-multiplatform.rb:28]: Can you propose a more substantial test than checking the version or calling for help? These tests rarely ever fail even for broken installations.

>We want tests that don’t require any user input and test the basic functionality of the application. For example `foo build-foo input.foo` is a good test and (despite their widespread use) `foo --version` and `foo --help` are bad tests. 

Ref. https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula

**Review by carlocab — COMMENTED** [review]
Thanks for the PR. A few comments/suggestions.

File: Formula/udp2raw-multiplatform.rb (hunk 2 above)
Comment (crab2313) [reply Formula/udp2raw-multiplatform.rb]: removed

File: Formula/udp2raw-multiplatform.rb (hunk 1 above)
Comment (crab2313) [reply Formula/udp2raw-multiplatform.rb]: removed

File: Formula/udp2raw-multiplatform.rb (hunk 3 above)
Comment (crab2313) [reply Formula/udp2raw-multiplatform.rb:12]: No. Before this I have a patch to solve it. But patch is not acceptable for CI checker.

File: Formula/udp2raw-multiplatform.rb (hunk 4 above)
Comment (crab2313) [reply Formula/udp2raw-multiplatform.rb:21]: There is no make install to do this.

File: Formula/udp2raw-multiplatform.rb (hunk 5 above)
Comment (crab2313) [reply Formula/udp2raw-multiplatform.rb:33]: udp2raw-multiplatform is a client only implementation of the original udp2raw for better multi-platform support. There is no way to properly test the client without launching a server first.

Comment (SMillerDev) [reply Formula/udp2raw-multiplatform.rb:33]: In the link it specifies:
> If the software cannot function without credentials or requires a virtual machine, docker instance, etc. to run, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected. This is preferred over mocking a dependency.

Which I think can easily apply here.

Comment (crab2313) [reply Formula/udp2raw-multiplatform.rb]: New test by checking the permission denied failure.

**Comment by crab2313** [comment]
@carlocab Hello, any pending reviews?

File: Formula/udp2raw-multiplatform.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    run [opt_bin/"udp2raw_mp", "--conf-file", etc/"udp2raw_client.conf"]
+    keep_alive true
+    require_root true
+    log_path var/"log/udp2raw.log"
+    error_log_path var/"log/udp2raw.log"
+  end
+
+  test do
+    assert_match(/.+SOCK_RAW allocation failed: Operation not permitted/,
+      shell_output("#{bin}/udp2raw_mp -c -r 127.0.0.1:50001 -l 127.0.0.1:50000 --log-level 1 --disable-color", 255))
```
Comment (SMillerDev) [inline Formula/udp2raw-multiplatform.rb:34]: Let's make sure we're not accidentally testing against something else.
```suggestion
      shell_output("#{bin}/udp2raw_mp -c -r 127.0.0.1:#{free_port} -l 127.0.0.1:#{free_port} --log-level 1 --disable-color", 255))
```

File: Formula/udp2raw-multiplatform.rb (hunk 4 above)
Comment (SMillerDev) [reply Formula/udp2raw-multiplatform.rb:21]: What is there a `make install` for?

File: Formula/udp2raw-multiplatform.rb [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  homepage "https://github.com/wangyu-/udp2raw-multiplatform"
+  url "https://github.com/wangyu-/udp2raw-multiplatform/archive/refs/tags/20210111.0.tar.gz"
+  sha256 "712ad3c79b6ef5bf106c615823d0b0b3865d1c957f9838cf05c23b7ac7024438"
+  license "MIT"
+
+  depends_on "libnet"
+  uses_from_macos "libpcap"
+
+  def install
+    ENV["OPT"] = Utils.safe_popen_read("libnet-config", "--cflags")
```
[the remark this answers is not in the document]
Comment (crab2313) [reply Formula/udp2raw-multiplatform.rb]: The makefile simply run `libnet-config --libs` and assume the libnet is installed in standard location. Which is insufficent to find the libnet headers.

File: Formula/udp2raw-multiplatform.rb (hunk 4 above)
Comment (crab2313) [reply Formula/udp2raw-multiplatform.rb:16]: The hand-written makefile do not support make install.

File: Formula/udp2raw-multiplatform.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (SMillerDev) [reply Formula/udp2raw-multiplatform.rb:12]: Can you make an issue for that upstream?

[the remark this answers is not in the document]
Comment (crab2313) [reply Formula/udp2raw-multiplatform.rb]: https://github.com/wangyu-/udp2raw-multiplatform/issues/60

**Review by MikeMcQuaid — APPROVED** [review]
Looks good to me, thanks @crab2313!

**Comment by wangyu-** [comment]
@crab2313  the repo https://github.com/wangyu-/udp2raw-multiplatform doesn't contain up-to-date code any more since 2021

the feature has been merged into https://github.com/wangyu-/udp2raw

**Comment by crab2313** [comment]
> @crab2313 the repo https://github.com/wangyu-/udp2raw-multiplatform doesn't contain up-to-date code any more since 2021
> 
> the feature has been merged into https://github.com/wangyu-/udp2raw

The wiki page and README in https://github.com/wangyu-/udp2raw guide the user to use https://github.com/wangyu-/udp2raw-multiplatform on MACOS and Windows.

**Comment by wangyu-** [comment]
@crab2313 the wiki https://github.com/wangyu-/udp2raw-multiplatform/wiki did mention the code is no longer updated . But I admit it's not conspicuous. I will make it more conspicuous.

I cannot archive https://github.com/wangyu-/udp2raw-multiplatform since I am stilling using it to hold binary release and discussion of windows/mac/bsd versions.

**Comment by wangyu-** [comment]
I have updated https://github.com/wangyu-/udp2raw-multiplatform, hopefully it's conspicuous enough.

**Pull request merged**